File: src/manifest/parser.zig
const std = @import("std");
const types = @import("../types.zig");
const toml = @import("../util/toml.zig");
const Manifest = @import("types.zig").Manifest;
const Project = @import("types.zig").Project;
const DependencyEntry = @import("types.zig").DependencyEntry;
const Workspace = @import("types.zig").Workspace;
const Scripts = @import("types.zig").Scripts;
const ScriptEntry = @import("types.zig").ScriptEntry;
const Security = @import("types.zig").Security;
const Build = @import("types.zig").Build;

pub const Error = error{
    MissingProjectName,
    MissingProjectVersion,
    UnknownSourceType,
    InvalidRiskLevel,
};

pub fn parse(allocator: std.mem.Allocator, content: []const u8) !Manifest {
    var doc = try toml.parse(allocator, content);
    defer doc.deinit(allocator);

    var manifest = Manifest{};

    if (doc.getTable("project")) |tbl| {
        manifest.project = try parseProject(tbl);
    }

    if (doc.getTable("deps")) |tbl| {
        const deps = try parseDeps(allocator, tbl);
        manifest.deps = deps;
    }

    if (doc.getTable("workspace")) |tbl| {
        manifest.workspace = try parseWorkspace(allocator, tbl);
    }

    if (doc.getTable("scripts")) |tbl| {
        manifest.scripts = try parseScripts(allocator, tbl);
    }

    if (doc.getTable("security")) |tbl| {
        manifest.security = try parseSecurity(tbl);
    }

    if (doc.getTable("build")) |tbl| {
        manifest.build = try parseBuild(tbl);
    }

    return manifest;
}

fn parseProject(table: *const toml.Table) !Project {
    var proj = Project{ .name = "", .version = "" };

    for (table.entries.items) |e| {
        if (std.mem.eql(u8, e.key, "name")) {
            proj.name = e.value.asString() orelse return error.MissingProjectName;
        } else if (std.mem.eql(u8, e.key, "version")) {
            proj.version = e.value.asString() orelse return error.MissingProjectVersion;
        } else if (std.mem.eql(u8, e.key, "description")) {
            proj.description = e.value.asString();
        } else if (std.mem.eql(u8, e.key, "license")) {
            proj.license = e.value.asString();
        } else if (std.mem.eql(u8, e.key, "repository")) {
            proj.repository = e.value.asString();
        } else if (std.mem.eql(u8, e.key, "homepage")) {
            proj.homepage = e.value.asString();
        }
    }

    return proj;
}

fn parseDeps(allocator: std.mem.Allocator, table: *const toml.Table) ![]DependencyEntry {
    var deps = std.ArrayList(DependencyEntry).init(allocator);

    for (table.entries.items) |e| {
        const inline_entries = switch (e.value) {
            .inline_table => |entries| entries,
            else => return error.UnknownSourceType,
        };

        var entry = DependencyEntry{
            .name = e.key,
            .source = .npm,
        };

        var has_source = false;
        for (inline_entries) |ie| {
            if (std.mem.eql(u8, ie.key, "source")) {
                const src_str = ie.value.asString() orelse return error.UnknownSourceType;
                entry.source = types.SourceType.fromString(src_str) catch return error.UnknownSourceType;
                has_source = true;
            } else if (std.mem.eql(u8, ie.key, "version")) {
                entry.version = ie.value.asString();
            } else if (std.mem.eql(u8, ie.key, "repo")) {
                entry.repo = ie.value.asString();
            } else if (std.mem.eql(u8, ie.key, "url")) {
                entry.url = ie.value.asString();
            } else if (std.mem.eql(u8, ie.key, "commit")) {
                entry.commit = ie.value.asString();
            } else if (std.mem.eql(u8, ie.key, "path")) {
                entry.path = ie.value.asString();
            } else if (std.mem.eql(u8, ie.key, "package")) {
                entry.package = ie.value.asString();
            }
        }

        if (!has_source) return error.UnknownSourceType;
        try deps.append(entry);
    }

    return deps.toOwnedSlice();
}

fn parseWorkspace(allocator: std.mem.Allocator, table: *const toml.Table) !Workspace {
    for (table.entries.items) |e| {
        if (std.mem.eql(u8, e.key, "members")) {
            const arr = switch (e.value) {
                .array => |items| items,
                else => return error.InvalidManifest,
            };
            const member_list = try allocator.alloc([]const u8, arr.len);
            for (arr, 0..) |v, i| {
                member_list[i] = switch (v) {
                    .string => |s| s,
                    else => return error.InvalidManifest,
                };
            }
            return Workspace{ .members = member_list };
        }
    }
    return Workspace{ .members = &.{} };
}

fn parseScripts(allocator: std.mem.Allocator, table: *const toml.Table) !Scripts {
    var script_list = std.ArrayList(ScriptEntry).init(allocator);

    for (table.entries.items) |e| {
        const cmd = e.value.asString() orelse continue;
        try script_list.append(.{ .name = e.key, .command = cmd });
    }

    return Scripts{ .entries = try script_list.toOwnedSlice() };
}

fn parseSecurity(table: *const toml.Table) !Security {
    var sec = Security{};

    for (table.entries.items) |e| {
        if (std.mem.eql(u8, e.key, "risk_threshold")) {
            const val = e.value.asString() orelse continue;
            sec.risk_threshold = std.meta.stringToEnum(types.RiskLevel, val);
        } else if (std.mem.eql(u8, e.key, "require_review")) {
            sec.require_review = e.value.asBoolean();
        } else if (std.mem.eql(u8, e.key, "allow_lifecycle_scripts")) {
            sec.allow_lifecycle_scripts = e.value.asBoolean();
        } else if (std.mem.eql(u8, e.key, "block_critical")) {
            sec.block_critical = e.value.asBoolean();
        }
    }

    return sec;
}

fn parseBuild(table: *const toml.Table) !Build {
    var b = Build{};

    for (table.entries.items) |e| {
        if (std.mem.eql(u8, e.key, "hermetic")) {
            b.hermetic = e.value.asBoolean();
        } else if (std.mem.eql(u8, e.key, "offline_first")) {
            b.offline_first = e.value.asBoolean();
        }
    }

    return b;
}

test "manifest: parse minimal" {
    const src =
        \\[project]
        \\name = "my-app"
        \\version = "0.1.0"
    ;
    var m = try parse(std.testing.allocator, src);
    defer m.deinit(std.testing.allocator);
    try std.testing.expectEqualStrings("my-app", m.project.name);
    try std.testing.expectEqualStrings("0.1.0", m.project.version);
}

test "manifest: parse with deps" {
    const src =
        \\[project]
        \\name = "app"
        \\version = "1.0.0"
        \\
        \\[deps]
        \\zod = { source = "npm", version = "3.23.8" }
        \\react = { source = "github", repo = "facebook/react", version = "18.x" }
    ;
    var m = try parse(std.testing.allocator, src);
    defer m.deinit(std.testing.allocator);
    try std.testing.expectEqual(@as(usize, 2), m.deps.len);
    try std.testing.expectEqualStrings("zod", m.deps[0].name);
    try std.testing.expectEqual(.npm, m.deps[0].source);
    try std.testing.expectEqualStrings("3.23.8", m.deps[0].version.?);
    try std.testing.expectEqualStrings("react", m.deps[1].name);
    try std.testing.expectEqualStrings("facebook/react", m.deps[1].repo.?);
}

test "manifest: parse workspace" {
    const src =
        \\[project]
        \\name = "monorepo"
        \\version = "0.1.0"
        \\
        \\[workspace]
        \\members = ["apps/*", "packages/*"]
    ;
    var m = try parse(std.testing.allocator, src);
    defer m.deinit(std.testing.allocator);
    try std.testing.expect(m.workspace != null);
    try std.testing.expectEqual(@as(usize, 2), m.workspace.?.members.len);
    try std.testing.expectEqualStrings("apps/*", m.workspace.?.members[0]);
}

test "manifest: parse security and build" {
    const src =
        \\[project]
        \\name = "secure-app"
        \\version = "0.1.0"
        \\
        \\[security]
        \\risk_threshold = "medium"
        \\require_review = true
        \\
        \\[build]
        \\hermetic = true
        \\offline_first = true
    ;
    var m = try parse(std.testing.allocator, src);
    defer m.deinit(std.testing.allocator);
    try std.testing.expectEqual(.medium, m.security.?.risk_threshold.?);
    try std.testing.expectEqual(true, m.security.?.require_review.?);
    try std.testing.expectEqual(true, m.build.?.hermetic.?);
    try std.testing.expectEqual(true, m.build.?.offline_first.?);
}

File: src/src/manifest/parser.rs
use std::collections::BTreeMap;

use crate::manifest::types::{
    Build, DepEntryRaw, DependencyEntry, Manifest, Project, ScriptEntry, Security, Workspace,
};

#[derive(Debug, thiserror::Error)]
pub enum ManifestParseError {
    #[error("missing project name")]
    MissingProjectName,
    #[error("missing project version")]
    MissingProjectVersion,
    #[error("unknown source type")]
    UnknownSourceType,
    #[error("invalid risk level")]
    InvalidRiskLevel,
    #[error("toml parse error: {0}")]
    Toml(#[from] toml::de::Error),
}

#[derive(serde::Deserialize)]
struct ManifestRaw {
    project: Option<ProjectRaw>,
    deps: Option<BTreeMap<String, DepEntryRaw>>,
    workspace: Option<WorkspaceRaw>,
    scripts: Option<BTreeMap<String, String>>,
    security: Option<SecurityRaw>,
    build: Option<BuildRaw>,
}

#[derive(serde::Deserialize)]
#[allow(dead_code)]
struct ProjectRaw {
    name: Option<String>,
    version: Option<String>,
    description: Option<String>,
    license: Option<String>,
    repository: Option<String>,
    homepage: Option<String>,
}

#[derive(serde::Deserialize)]
struct WorkspaceRaw {
    members: Option<Vec<String>>,
}

#[derive(serde::Deserialize)]
struct SecurityRaw {
    risk_threshold: Option<String>,
    require_review: Option<bool>,
}

#[derive(serde::Deserialize)]
struct BuildRaw {
    hermetic: Option<bool>,
    offline_first: Option<bool>,
}

pub fn parse(content: &str) -> Result<Manifest, ManifestParseError> {
    let raw: ManifestRaw = toml::from_str(content)?;

    let project = match raw.project {
        Some(p) => {
            let name = p.name.unwrap_or_default();
            let version = p.version.unwrap_or_default();
            if name.is_empty() {
                return Err(ManifestParseError::MissingProjectName);
            }
            if version.is_empty() {
                return Err(ManifestParseError::MissingProjectVersion);
            }
            Project {
                name,
                version,
            }
        }
        None => return Err(ManifestParseError::MissingProjectName),
    };

    let valid_sources = ["npm", "registry", "github", "git", "local", "workspace"];

    let deps = match raw.deps {
        Some(map) => map
            .into_iter()
            .map(|(name, raw)| {
                let source = raw.source.unwrap_or_else(|| "npm".to_string());
                if !valid_sources.contains(&source.as_str()) {
                    return Err(ManifestParseError::UnknownSourceType);
                }
                Ok(DependencyEntry {
                    name,
                    source,
                    version: raw.version,
                    repo: raw.repo,
                    url: raw.url,
                    commit: raw.commit,
                    path: raw.path,
                })
            })
            .collect::<Result<Vec<_>, _>>()?,
        None => Vec::new(),
    };

    let valid_risk_levels = ["low", "medium", "high", "critical"];

    let security = raw
        .security
        .map(|s| {
            if let Some(ref threshold) = s.risk_threshold {
                if !valid_risk_levels.contains(&threshold.as_str()) {
                    return Err(ManifestParseError::InvalidRiskLevel);
                }
            }
            Ok(Security {
                risk_threshold: s.risk_threshold,
                require_review: s.require_review,
            })
        })
        .transpose()?;

    let workspace = raw
        .workspace
        .and_then(|w| w.members.map(|members| Workspace { members }));

    let scripts = raw.scripts.map_or_else(Vec::new, |map| {
        map.into_iter()
            .map(|(name, command)| ScriptEntry { name, command })
            .collect()
    });

    let build = raw.build.map(|b| Build {
        hermetic: b.hermetic,
        offline_first: b.offline_first,
    });

    Ok(Manifest {
        project,
        deps,
        workspace,
        scripts,
        security,
        build,
    })
}

#[cfg(test)]
mod tests {
    #![allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]
    use super::*;

    #[test]
    fn test_parse_minimal() {
        let src = r#"
            [project]
            name = "my-app"
            version = "0.1.0"
        "#;
        let m = parse(src).unwrap();
        assert_eq!(m.project.name, "my-app");
        assert_eq!(m.project.version, "0.1.0");
    }

    #[test]
    fn test_parse_with_deps() {
        let src = r#"
            [project]
            name = "app"
            version = "1.0.0"

            [deps]
            zod = { source = "npm", version = "3.23.8" }
            react = { source = "github", repo = "facebook/react", version = "18.x" }
        "#;
        let m = parse(src).unwrap();
        assert_eq!(m.deps.len(), 2);
        let zod = m.deps.iter().find(|d| d.name == "zod").unwrap();
        assert_eq!(zod.source, "npm");
        assert_eq!(zod.version.as_deref(), Some("3.23.8"));
        let react = m.deps.iter().find(|d| d.name == "react").unwrap();
        assert_eq!(react.repo.as_deref(), Some("facebook/react"));
    }

    #[test]
    fn test_parse_workspace() {
        let src = r#"
            [project]
            name = "monorepo"
            version = "0.1.0"

            [workspace]
            members = ["apps/*", "packages/*"]
        "#;
        let m = parse(src).unwrap();
        assert!(m.workspace.is_some());
        assert_eq!(m.workspace.as_ref().unwrap().members.len(), 2);
        assert_eq!(m.workspace.as_ref().unwrap().members[0], "apps/*");
    }

    #[test]
    fn test_parse_missing_project_section() {
        let src = r#"
            name = "no-project"
            version = "0.1.0"
        "#;
        match parse(src) {
            Err(ManifestParseError::MissingProjectName) => {}
            other => panic!("expected MissingProjectName, got {other:?}"),
        }
    }

    #[test]
    fn test_parse_missing_project_version() {
        let src = r#"
            [project]
            name = "app"
        "#;
        match parse(src) {
            Err(ManifestParseError::MissingProjectVersion) => {}
            other => panic!("expected MissingProjectVersion, got {other:?}"),
        }
    }

    #[test]
    fn test_parse_unknown_source_type() {
        let src = r#"
            [project]
            name = "app"
            version = "0.1.0"

            [deps]
            foo = { source = "nonexistent", version = "1.0.0" }
        "#;
        match parse(src) {
            Err(ManifestParseError::UnknownSourceType) => {}
            other => panic!("expected UnknownSourceType, got {other:?}"),
        }
    }

    #[test]
    fn test_parse_invalid_risk_level() {
        let src = r#"
            [project]
            name = "app"
            version = "0.1.0"

            [security]
            risk_threshold = "bogus"
        "#;
        match parse(src) {
            Err(ManifestParseError::InvalidRiskLevel) => {}
            other => panic!("expected InvalidRiskLevel, got {other:?}"),
        }
    }

    #[test]
    fn test_parse_security_and_build() {
        let src = r#"
            [project]
            name = "secure-app"
            version = "0.1.0"

            [security]
            risk_threshold = "medium"
            require_review = true

            [build]
            hermetic = true
            offline_first = true
        "#;
        let m = parse(src).unwrap();
        assert_eq!(
            m.security.as_ref().unwrap().risk_threshold.as_deref(),
            Some("medium")
        );
        assert_eq!(m.security.as_ref().unwrap().require_review, Some(true));
        assert_eq!(m.build.as_ref().unwrap().hermetic, Some(true));
        assert_eq!(m.build.as_ref().unwrap().offline_first, Some(true));
    }

    #[test]
    fn test_parse_scripts() {
        let src = r#"
            [project]
            name = "app"
            version = "0.1.0"

            [scripts]
            build = "tsc"
            test = "vitest"
        "#;
        let m = parse(src).unwrap();
        assert_eq!(m.scripts.len(), 2);
        assert_eq!(m.scripts[0].name, "build");
        assert_eq!(m.scripts[0].command, "tsc");
        assert_eq!(m.scripts[1].name, "test");
        assert_eq!(m.scripts[1].command, "vitest");
    }

    #[test]
    fn test_parse_unicode_name() {
        let src = r#"
            [project]
            name = "🔥-test-中文"
            version = "0.1.0"
        "#;
        let m = parse(src).unwrap();
        assert_eq!(m.project.name, "🔥-test-中文");
    }

    #[test]
    fn test_parse_long_name() {
        let name = "a".repeat(10_000);
        let src = format!(
            r#"
            [project]
            name = "{name}"
            version = "0.1.0"
        "#
        );
        let m = parse(&src).unwrap();
        assert_eq!(m.project.name.len(), 10_000);
    }

    #[test]
    fn test_parse_deeply_nested_table() {
        let src = r#"
            [project]
            name = "weird"
            version = "0.1.0"

            [deps]
            pkg = { source = "npm", version = "1.0.0" }

            [a.b.c.d.e.f.g]
            x = 1
        "#;
        parse(src).unwrap();
    }

    #[test]
    fn test_parse_name_with_path_chars() {
        let src = r#"
            [project]
            name = "../../etc/passwd"
            version = "0.1.0"
        "#;
        let m = parse(src).unwrap();
        assert_eq!(m.project.name, "../../etc/passwd");
    }

    #[test]
    fn test_parse_huge_integer() {
        let src = r#"
            [project]
            name = "bigint"
            version = "0.1.0"

            [security]
            require_review = true
        "#;
        // TOML integers beyond i64 range cause parse error
        let huge = format!("{src}\nhuge = 999999999999999999999999999");
        let result = parse(&huge);
        assert!(result.is_err());
    }
}
